File: front/frontend-admin/app/api/ErrorServiceApi.ts
/// <reference path="api.d.ts" />

/* tslint:disable:no-unused-variable member-ordering */

namespace api {
    'use strict';

    export class ErrorServiceApi {
        protected basePath = '.';
        public defaultHeaders : any = {};

        static $inject: string[] = ['$http', '$httpParamSerializer'];

        constructor(protected $http: ng.IHttpService, protected $httpParamSerializer?: (d: any) => any, basePath?: string) {
            if (basePath) {
                this.basePath = basePath;
            }
        }

        private extendObj<T1,T2>(objA: T1, objB: T2) {
            for(let key in objB){
                if(objB.hasOwnProperty(key)){
                    objA[key] = objB[key];
                }
            }
            return <T1&T2>objA;
        }

        /**
         * Get the list of errors sorted by date desc
         * 
         * @param select select
         * @param page page
         * @param size size
         */
        public getErrors (select: Object, page?: number, size?: number, extraHttpRequestParams?: any ) : ng.IHttpPromise<PageOfErrorMessage> {
            const localVarPath = this.basePath + '/api/services/error';

            let queryParameters: any = {};
            let headerParams: any = this.extendObj({}, this.defaultHeaders);
            // verify required parameter 'select' is set
            if (!select) {
                throw new Error('Missing required parameter select when calling getErrors');
            }
            if (select !== undefined) {
                queryParameters['select'] = select;
            }

            if (page !== undefined) {
                queryParameters['page'] = page;
            }

            if (size !== undefined) {
                queryParameters['size'] = size;
            }

            let httpRequestParams: any = {
                method: 'GET',
                url: localVarPath,
                json: true,
                
                
                params: queryParameters,
                headers: headerParams
            };

            if (extraHttpRequestParams) {
                httpRequestParams = this.extendObj(httpRequestParams, extraHttpRequestParams);
            }

            return this.$http(httpRequestParams);
        }
        /**
         * Ignoring a selection of errors
         * Warning! This could be harmful for your workflow
         * @param select select
         */
        public dismiss (select: ErrorSelector, extraHttpRequestParams?: any ) : ng.IHttpPromise<number> {
            const localVarPath = this.basePath + '/api/services/error/ignore';

            let queryParameters: any = {};
            let headerParams: any = this.extendObj({}, this.defaultHeaders);
            // verify required parameter 'select' is set
            if (!select) {
                throw new Error('Missing required parameter select when calling dismiss');
            }
            let httpRequestParams: any = {
                method: 'POST',
                url: localVarPath,
                json: true,
                data: select,
                
                
                params: queryParameters,
                headers: headerParams
            };

            if (extraHttpRequestParams) {
                httpRequestParams = this.extendObj(httpRequestParams, extraHttpRequestParams);
            }

            return this.$http(httpRequestParams);
        }
        /**
         * Recover a selection of errors
         * 
         * @param select select
         */
        public recover (select: ErrorSelector, extraHttpRequestParams?: any ) : ng.IHttpPromise<number> {
            const localVarPath = this.basePath + '/api/services/error/recover';

            let queryParameters: any = {};
            let headerParams: any = this.extendObj({}, this.defaultHeaders);
            // verify required parameter 'select' is set
            if (!select) {
                throw new Error('Missing required parameter select when calling recover');
            }
            let httpRequestParams: any = {
                method: 'POST',
                url: localVarPath,
                json: true,
                data: select,
                
                
                params: queryParameters,
                headers: headerParams
            };

            if (extraHttpRequestParams) {
                httpRequestParams = this.extendObj(httpRequestParams, extraHttpRequestParams);
            }

            return this.$http(httpRequestParams);
        }
        /**
         * Get a specific error
         * 
         * @param id id
         */
        public getError (id: string, extraHttpRequestParams?: any ) : ng.IHttpPromise<ErrorMessage> {
            const localVarPath = this.basePath + '/api/services/error/{id}'
                .replace('{' + 'id' + '}', String(id));

            let queryParameters: any = {};
            let headerParams: any = this.extendObj({}, this.defaultHeaders);
            // verify required parameter 'id' is set
            if (!id) {
                throw new Error('Missing required parameter id when calling getError');
            }
            let httpRequestParams: any = {
                method: 'GET',
                url: localVarPath,
                json: true,
                
                
                params: queryParameters,
                headers: headerParams
            };

            if (extraHttpRequestParams) {
                httpRequestParams = this.extendObj(httpRequestParams, extraHttpRequestParams);
            }

            return this.$http(httpRequestParams);
        }
    }
}

File: backend/workflow/src/main/java/eu/researchalps/api/ErrorServiceApi.java
package eu.researchalps.api;

import com.datapublica.companies.api.selector.ErrorSelector;
import com.datapublica.companies.model.error.ErrorMessage;
import com.datapublica.companies.repository.mongo.ErrorRepository;
import com.datapublica.companies.workflow.service.ErrorHandler;
import eu.researchalps.api.exception.NotFoundException;
import eu.researchalps.api.util.ApiConstants;
import io.swagger.annotations.ApiOperation;
import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

/**
 *
 */
@Controller
@RequestMapping("/services/error")
public class ErrorServiceApi {
    @Autowired
    private ErrorHandler errorHandler;
    @Autowired
    private ErrorRepository errorRepository;

    @Autowired
    private ErrorRepository errors;

    /**
     * Find errors using an error selector.
     *
     * @param query The error selector
     * @param page The page number
     * @param size The page size
     * @return The matched errors
     */
    @ResponseBody
    @RequestMapping(method = RequestMethod.GET, produces = ApiConstants.PRODUCES_JSON)
    @ApiOperation("Get the list of errors sorted by date desc")
    public Page<ErrorMessage> getErrors(@RequestParam("select") ErrorSelector query, @RequestParam(value = "page", required = false, defaultValue = "1") int page, @RequestParam(value = "size", required = false, defaultValue = "100") int size) {
        return this.errors.select(query, new PageRequest(page - 1, size));
    }

    /**
     * Find an error by id.
     *
     * @param id The error id
     * @return The corresponding error
     */
    @ResponseBody
    @RequestMapping(value = "/{id}", method = RequestMethod.GET, produces = ApiConstants.PRODUCES_JSON)
    @ApiOperation("Get a specific error")
    public ErrorMessage getError(@PathVariable("id") String id) throws NotFoundException {
        final ErrorMessage error = this.errors.findOne(new ObjectId(id));
        if(error != null) {
            return error;
        }
        throw new NotFoundException("error", id);
    }

    @ResponseBody
    @RequestMapping(value = "/recover", method = RequestMethod.POST)
    @ApiOperation("Recover a selection of errors")
    public long recover(@RequestBody ErrorSelector select) {
        return errorRepository.select(select).filter(errorHandler::recover).count();
    }

    @ResponseBody
    @RequestMapping(value = "/ignore", method = RequestMethod.POST)
    @ApiOperation(value = "Ignoring a selection of errors", notes = "Warning! This could be harmful for your workflow")
    public long dismiss(@RequestBody ErrorSelector select) {
        return  errorRepository.select(select).peek(errorRepository::delete).count();
    }
}
